coins/bitcoin/src/crypto.rs

```diff
@@ -0,0 +1,135 @@
+};
+
+pub fn make_even(mut key: ProjectivePoint) -> (ProjectivePoint, u64) {
+  let mut c = 0;
+  while key.to_encoded_point(true).tag() == Tag::CompressedOddY {
+    key += ProjectivePoint::GENERATOR;
+    c += 1;
+  }
+  (key, c)
+}
+
+#[derive(Clone)]
+pub struct BitcoinHram {}
+
+impl Hram<Secp256k1> for BitcoinHram {
+  fn hram(R: &ProjectivePoint, A: &ProjectivePoint, m: &[u8]) -> Scalar {
+    let (R, _) = make_even(*R);
+
+    let r_encoded_point = R.to_encoded_point(true);
+    let a_encoded_point = A.to_encoded_point(true);
+    let tag = b"BIP0340/challenge";
+    let tag_hash = Sha256::digest(tag);
+    let mut data = Sha256::new();
+    data.update(tag_hash);
+    data.update(tag_hash);
+    data.update(r_encoded_point.x().to_owned().unwrap());
+    data.update(a_encoded_point.x().to_owned().unwrap());
+    data.update(&m[..]);
+
+    Scalar::from_uint_reduced(U256::from_be_slice(&data.finalize()))
+  }
+}
+
+/// Signing error
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum SignerError {
+  /// The private key is missing for the required public key
+  MissingKey,
+  /// The private key in use has the right fingerprint but derives differently than expected
+  InvalidKey,
+  /// The user canceled the operation
+  UserCanceled,
+  /// Input index is out of range
+  InputIndexOutOfRange,
+  /// The `non_witness_utxo` field of the transaction is required to sign this input
+  MissingNonWitnessUtxo,
+  /// The `non_witness_utxo` specified is invalid
+  InvalidNonWitnessUtxo,
+  /// The `witness_utxo` field of the transaction is required to sign this input
+  MissingWitnessUtxo,
+  /// The `witness_script` field of the transaction is required to sign this input
+  MissingWitnessScript,
+  /// The fingerprint and derivation path are missing from the psbt input
+  MissingHdKeypath,
+  /// The psbt contains a non-`SIGHASH_ALL` sighash in one of its input and the user hasn't
+  /// explicitly allowed them
+  ///
+  /// To enable signing transactions with non-standard sighashes set
+  /// [`SignOptions::allow_all_sighashes`] to `true`.
+  NonStandardSighash,
+  /// Invalid SIGHASH for the signing context in use
+  InvalidSighash,
+  /// Error while computing the hash to sign
+  SighashError(sighash::Error),
+  /// Error while signing using hardware wallets
+  #[cfg(feature = "hardware-signer")]
+  HWIError(hwi::error::Error),
+}
+
+pub trait PSBTUtils {
+  fn get_utxo_for(&self, input_index: usize) -> Option<TxOut>;
+}
+
+impl PSBTUtils for PartiallySignedTransaction {
+  fn get_utxo_for(&self, input_index: usize) -> Option<TxOut> {
+    let tx = &self.unsigned_tx;
+
+    if input_index >= tx.input.len() {
+      return None;
+    }
+
+    if let Some(input) = self.inputs.get(input_index) {
+      if let Some(wit_utxo) = &input.witness_utxo {
+        Some(wit_utxo.clone())
+      } else if let Some(in_tx) = &input.non_witness_utxo {
+        Some(in_tx.output[tx.input[input_index].previous_output.vout as usize].clone())
```

Comment:
Please use the explicit `usize::try_from(tx.input[input_index].previous_output.vout).unwrap()`.